fix: generated python with ifabsent

Silvanoc opened the pull request:
Generated Python code when `ifabsent` specifies an enumeration permissible value was erroneous. This patch fixes it.

Fixes #2284

Silvanoc commented:
Regression test added reproducing the issue with commit d576720b65cdc55bc659428c9655ff63850ea546 see [failing run](https://github.com/linkml/linkml/actions/runs/11071303849) before pushing the fix.

Silvanoc commented:
@pkalita-lbl I am leaving this PR as Draft because I am not 100% sure that the fix is correct.
Although the tests are passing and I get the expected output. Please take a detailed look at it and let me know if you find the fix is correct (to mark the PR as ready) or not (in this case please help me to fix it).

pkalita-lbl commented on linkml/generators/pythongen.py:
```diff
@@ -131,6 +131,14 @@ def gen_schema(self) -> str:
+                enum_pvs.append(f"{enum_name}({pv})")
+        for class_name, class_def in self.schema.classes.items():
+            for slot in class_def.slots:
+                if self.schema.slots[slot].ifabsent in enum_pvs:
+                    enumimports = enumimports + "from dataclasses import field\n"
```
If multiple classes/slots satisfy this condition, I think this would result in multiple identical imports?

sneakers-the-rat replied:
pydanticgen `Import` and `Imports` classes handle deduplication and sorting of imports (and are not pydantic specific) https://github.com/linkml/linkml/blob/68a880a458708c20cde6337b3493ee93a85c3ba8/linkml/generators/pydanticgen/template.py#L263

Silvanoc replied:
I have used the `Import` and `Imports` classes as proposed by @sneakers-the-rat. They are really cool!

I have not changed the rest of the imports to keep this PR scoped. But IMO a future PR should take care of using those same classes all the imports in the generated python code.

Silvanoc replied:
BTW @pkalita-lbl good catch!

pkalita-lbl commented on linkml/generators/python/python_ifabsent_processor.py:
```diff
@@ -71,7 +71,7 @@ def map_uri_default_value(self, default_value: str, slot: SlotDefinition, cls: C
         self, enum_name: EnumDefinitionName, permissible_value_name: str, slot: SlotDefinition, cls: ClassDefinition
     ):
-        return f"{enum_name}.{permissible_value_name}"
+        return f"field(default_factory={enum_name}.{permissible_value_name})"
```
I'm not quite sure I understand the need for using `field` with `default_factory` here. Can you elaborate?

sneakers-the-rat replied:
this would break pydanticgen (uses `Field` not `field`) and throws an error

```python
from dataclasses import dataclass, field
from enum import Enum
class MyEnum(Enum):
    value = 'value'
    
@dataclass
class MyClass():
    afield: MyEnum = field(default_factory=MyEnum.value)
    
MyClass()
Traceback (most recent call last):
  File "/Users/jonny/Library/Caches/pypoetry/virtualenvs/linkml-G-1YJs6r-py3.11/lib/python3.11/site-packages/IPython/core/interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-14-a39b5d23b52a>", line 1, in <module>
    MyClass()
  File "<string>", line 3, in __init__
TypeError: 'MyEnum' object is not callable
```

Silvanoc replied:
I am not an expert in python dataclasses, therefore I haven't found a better fix for following error that I get if `field(default_factory())` is not used.
``` shell
ValueError: mutable default <class 'linkml_runtime.linkml_model.meta.PermissibleValue'> for field ifabsent_not_literal is not allowed: use default_factory
```

You might know a better way to fix that issue without breaking pydanticgen.

BTW @sneakers-the-rat tests are passing, are we missing some tests on pydanticgen? Or how did you catch it?

pkalita-lbl reviewed (commented):
Thanks @Silvanoc! I think having the enum definitions before the class definitions makes sense. A couple questions about the other changes.

Silvanoc commented:
I am trying to declare a "derived" meta-model (one derived from the LinkML meta-model) and I am stumbling upon the same issue in `shaclgen`.

I don't really understand why the range of the slot `range` is `enum_definition`, in any case I am getting an error again because `ifabsent` is failing on resolving the default value. I have not tried yet, if my patch resolves `shaclgen` too.

Silvanoc commented:
> I am trying to declare a "derived" meta-model (one derived from the LinkML meta-model) and I am stumbling upon the same issue in `shaclgen`.
> 
> I don't really understand why the range of the slot `range` is `enum_definition`, in any case I am getting an error again because `ifabsent` is failing on resolving the default value. I have not tried yet, if my patch resolves `shaclgen` too.

This issue can be also easily reproduced installing the latest LinkML version (v1.8.4), getting the latest LinkML meta-model and running `gen-shacl linkml_model/model/schema/meta.yaml`.

This is the error:
``` shell
ValueError: The ifabsent value `default_range` of the `range` slot could not be processed
```

I have run it with the debugger and the issue has the same cause: `ifabsent` failing on enumerations.

I have also applied my patch of this branch and it does not seem to fix it 😞 We probably need a better analysis of someone with deeper knowledge than me on the LinkML codebase and Python programming. @vincentkelleher since you wrote the new ifabsent code, you might be the right guy for this...

IMO this is a "hidden" issue that might have deeper consequences than thought!

Silvanoc commented:
I have taken my time to go the issue to its root-cause and I think I got it.

I do not know how test the generated Pydantic code like @sneakers-the-rat did for the 1st try. I have tried against the [kitchen_sink_ifabsent.yaml](https://github.com/linkml/linkml/blob/main/tests/test_generators/input/kitchen_sink_ifabsent.yaml) schema. But then I have realized that I get exactly the same error that you get with the current `main` branch. So it is not an issue I have introduced 😄 

I have also replaced the generation of the imports to use the same module being used by pydanticgen. The result would make `isort` proud...

I would recommend reviewing on the individual commits.

sneakers-the-rat commented:
Oh boy if you swapped the import system ill gladly review.

sierra-moxon commented:
Thank you @Silvanoc!  And for bringing this one to our attention on slack.  It is indeed breaking other model distributions as well.

sneakers-the-rat commented on linkml/generators/pythongen.py:
```diff
@@ -126,14 +127,135 @@ def visit_type(self, typ: TypeDefinition) -> None:
-            ""
-            if self.genmeta
-            else "from linkml_runtime.linkml_model.meta import EnumDefinition, PermissibleValue, PvFormulaOptions\n"
+        if not self.genmeta:
```
eg. could do this

at a module level...

```python
DEFAULT_IMPORTS = Imports(imports=[
    Import(module="dataclasses", objects=[ObjectImport(name="dataclass")]),
    # ... the rest of the default imports
])

METAMODEL_IMPORTS = Imports(imports=[
    Import(module="linkml_runtime.linkml_model.meta", objects=[ObjectImport(name="EnumDefinition")]
    # ... etc.
])

# and any other import groups
```

and then in the generator you could do

```python
def gen_schema(self):
    imports = DEFAULT_IMPORTS.model_copy()
    if not self.genmeta:
        imports += METAMODEL_IMPORTS
```

Silvanoc replied:
Fully agree here too. But refactoring that I would rather address in a separate PR to bring this one to an end.

sneakers-the-rat replied:
i think it would take approximately 10 seconds, but i don't use the python generator so it doesn't really matter that much to me

Silvanoc replied:
My experience is that you know when you start refactoring, but you never know when you will end 😉

And I have the feeling that my PRs tend to get much too big and extensive over time to be digestible (see also #2303, with 104 comments!). That is why I am trying to close this one before it gets into another [Berlin Brandenburg Airport](https://en.wikipedia.org/wiki/Construction_of_Berlin_Brandenburg_Airport).

sneakers-the-rat replied:
sure sure, do whatever, like i said i don't use the python generator so idrc <3

Silvanoc replied:
This should be also addresses on #2367.

sneakers-the-rat commented on linkml/generators/pythongen.py:
```diff
@@ -126,14 +127,135 @@ def visit_type(self, typ: TypeDefinition) -> None:
-        handlerimport = "from linkml_runtime.utils.enumerations import EnumDefinitionImpl"
+
+        if self._import_field():
+            all_imports = all_imports + Import(module="dataclasses", objects=[ObjectImport(name="field")])
```
might as well just always import this, since importing one object from a module means executing the whole module, this wouldn't be a perf hit. pydanticgen uses a system of build results to handle imports that are needed in specific cases and not others, but retrofitting that into the pythongen would be a lot of work.

Silvanoc replied:
I agree that the complexity of `_import_field` might not be worth. You will only get linter warnings on the generated code, but you get anyway so many other warnings 😄 I will change it as you propose.

vincentkelleher replied:
I like @Silvanoc's approach on generating only the necessary imports but each point of view has its pros and cons 😇

sneakers-the-rat replied:
i mean yeah i designed a whole system of build result models with the primary purpose of only including necessary imports for pydanticgen, so of course i prefer doing that, just saying that gating with private methods just doesn't scale that well

Silvanoc replied:
Although I appreciate @vincentkelleher support on my approach. I have to agree with @sneakers-the-rat that saving that `field` import is not worth the added complexity. So I will go for the always import `field` approach.

sneakers-the-rat replied:
don't mean to be a downer, also am indifferent on this change, defer to the author :)

Silvanoc replied:
I have revised my decision. Although I am not happy with the complexity added with `_import_field`, importing always `dataclasses.field` results in a cascade of snapshots needing updating. I have decided to leave `_import_field` there... At least for the time being. I will consider afterwards removing it.

sneakers-the-rat commented on linkml/generators/pythongen.py:
```diff
@@ -841,7 +949,13 @@ def gen_postinit(self, cls: ClassDefinition, slot: SlotDefinition) -> Optional[s
-                    rlines.append(f"\tself.{aliased_slot_name} = {base_type_name}(self.{aliased_slot_name})")
+                    enum_forward_ref = self.ifabsent_processor.enum_forward_references
+                    if cls.name in enum_forward_ref.keys() and slot.range in enum_forward_ref[cls.name].keys():
+                        rlines.append(
```
sort of having a hard time telling what's happening here, but i'll see if running it and inspecting the output clears things up

Silvanoc replied:
This is probably the trickiest part of the whole patch. It is probably worth a more extensive comment explaining it.

vincentkelleher replied:
From what I understand, in the case of a forward reference for an enum it will return something like:

```python
self.my_slot = MyType.MyEnum(MyPermissibleValue)
```

And in the case of a basic type field, it will return:

```python
self.my_slot = MyType(self.my_slot)
```

Did I understand that correctly ?

Silvanoc replied:
@vincentkelleher you are right, it should be `MyEnum.MyPermissibleValue`, I will check if it returns it or something wrong (like you seem to expect from my code).

Silvanoc replied:
The code was in fact buggy, but not as @vincentkelleher thought.

And this confirms that @sneakers-the-rat was right to "complain" about the lack of a specific test
> i can't see a test for the underlying problem (an ifabsent value that is a value of an enum) here

My poor-man regression test `test_ifabsent_stuff` was only asserting that the module could be generated and compiled, not that the resulting code was the expected one... I am almost done with the bug fixing and the corresponding test for `ifabsent` on enumerations.

Silvanoc replied:
Now `test_ifabsent_stuff` also tests the correctness of the `ifabsent` in the case of enumerations.

sneakers-the-rat reviewed (commented):
So I am not sure of the status of this PR, if it's done or not, but i can't see a test for the underlying problem (an `ifabsent` value that is a value of an enum) here? the tests i do see check for the existence of something in the `enum_forward_references` dict, but not that the default result is correctly set in an instance of that model? i might be missing something. 

I don't mean to throw cold water on all the work that you've done here (i love converting the import system), but I think that the easiest fix here probably would have been to just put enums before classes in the rendered module, since as far as i know enums can't refer to other objects that need to be defined first but classes can.

sneakers-the-rat commented on tests/test_generators/test_pythongen.py:
```diff
@@ -68,6 +68,10 @@ def test_multiline_stuff(input_path):
     )
 
 
+def test_ifabsent_stuff(input_path):
```
sorry should have added this to the previous review - what does this do?

Silvanoc replied:
It simply tries to generate the Python code corresponding to the schema and it fails without the patch. It is a sort of "poor-man" regression test.

sneakers-the-rat commented on tests/test_generators/python/test_python_ifabsent_processor.py:
```diff
@@ -403,9 +403,13 @@ def test_enum_default_value():
+        == "field(default=None)"
     )
 
+    print(processor.enum_forward_references)
```
no print statements plz :)

Silvanoc replied:
Ups, late-hours slip. Will remove it. Thanks for the good catch!

sneakers-the-rat reviewed (commented):
sorry for the double review, pressed submit before i was done. not sure if there's a way to add more comments to a previous review but here we are

sneakers-the-rat commented:
> This is the error:
> 
> ValueError: The ifabsent value `default_range` of the `range` slot could not be processed
> 
> I have run it with the debugger and the issue has the same cause: ifabsent failing on enumerations.

I think this is a separate issue than the order of definition problem re: enums and classes in the pythongen, basically `default_range` is a special case in `ifabsent` that is currently [explicitly skipped](https://github.com/linkml/linkml/blame/c4379e5a121732b6adda21284bb34dde590743b1/linkml/generators/python/python_ifabsent_processor.py#L14) by the python ifabsent processor and unhandled by the shaclgen. would recommend we split that off into a separate issue/pr

sneakers-the-rat commented:
here i just went and fixed the other issue with `ifabsent` for the special cases over here: https://github.com/linkml/linkml/pull/2366

Silvanoc commented on linkml/generators/pythongen.py:
```diff
@@ -126,14 +127,135 @@ def visit_type(self, typ: TypeDefinition) -> None:
+        return False
+
     def gen_schema(self) -> str:
+        all_imports = Imports()
```
Sure, but we would need a different "default", since generated Python code don't need to import Pydantic (for example).

I would like to focus here on fixing #2284 and address this proposal in a separate PR, since it is only a refactoring (no change on external behavior).

sneakers-the-rat replied:
well, sure, of course it would be different lol. i'm just saying cut and paste this block outside of the method and put it at the module level, but whatever you do u

Silvanoc replied:
Created #2367 to address this one.

Silvanoc commented:
I think I have addressed all open points and since this PR seems to be [breaking other schemas](https://github.com/linkml/linkml/pull/2329#issuecomment-2406185641), I am triggering all previous reviewers to have a last look at it.

I have also assigned this PR (apparently that is not the normal workflow in the project) just to signal that I assume that @pkalita-lbl will merge or reassign to someone "permitted" to merge once happy with the reached status.

Silvanoc commented:
I have been able to simplify the code even more by using a string as default for the field instead of using `field(default=None)`. I can also convey the information about which is the permissible value to be taken by default in that string. The `__post_init__` only needs to convert that string into a call to the real permissible value.

sneakers-the-rat commented on tests/test_generators/python/test_python_ifabsent_processor.py:
```diff
@@ -374,6 +374,44 @@ def test_uri_default_value():
+# PythonIfAbsentProcessor instance suddenly has some attributes
+# coming from tests using the kitchen-sink example! Invalidating that
+# way the results of the test.
+@pytest.fixture(scope="module")
```
this is a test, right? so it shouldn't be a fixture at all? I don't see how it interacts with the kitchen sink example here.

in any case it being a fixture meant it is not run as a test, note its absence (i did check out this branch and am looking at the code being there and it's in editable mode so i can cause tests to fail by throwing a spurious exception, so it's not my environment, just how fixtures work.

```
collected 15 items                                                                                            

tests/test_generators/python/test_python_ifabsent_processor.py::test_unimplemented_default_value[class_curie] PASSED [  6%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_unimplemented_default_value[class_uri] PASSED [ 13%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_unimplemented_default_value[slot_uri] PASSED [ 20%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_unimplemented_default_value[slot_curie] PASSED [ 26%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_unimplemented_default_value[default_range] PASSED [ 33%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_unimplemented_default_value[default_ns] PASSED [ 40%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_string_default_value PASSED        [ 46%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_boolean_default_value PASSED       [ 53%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_numeric_default_value PASSED       [ 60%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_time_default_value PASSED          [ 66%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_date_default_value PASSED          [ 73%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_datetime_default_value PASSED      [ 80%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_uri_default_value PASSED           [ 86%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_enum_default_value PASSED          [ 93%]
tests/test_generators/python/test_python_ifabsent_processor.py::test_bnode_default_value PASSED         [100%]

============================================= 15 passed in 0.77s ==============================================
```

sneakers-the-rat replied:
If i add a `pdb.set_trace()` call below the assignment and print it, i get what would be expected:

```yaml

id: ifabsent_tests
name: ifabsent_tests

prefixes:
  ex: https://example.org/
default_prefix: ex

classes:
  Student:
    attributes:

      - name: presence
        range: PresenceEnum
      - name: invalidPresence
        range: PresenceEnum

enums:
  PresenceEnum:
    permissible_values:
      Present:
        description: It's there.
      Missing:
        description: It's not there.
```

Silvanoc replied:
This was a desperate late-hours try to fix an issue that was driving me nuts without a logical explanation I could find.  Apparently my laptop only needed a nap, since I cannot reproduce the issue anymore 🤷

sneakers-the-rat commented on tests/test_generators/python/test_python_ifabsent_processor.py:
```diff
@@ -403,7 +441,7 @@ def test_enum_default_value():
             schema_view.all_classes()[ClassDefinitionName("Student")],
         )
-        == "PresenceEnum.Missing"
+        == "'PresenceEnum.Missing'"
```
Great! I like the string approach more than putting `None` there :) and testing this out it looks like it's working.

sneakers-the-rat commented on tests/test_generators/test_pythongen.py:
```diff
@@ -68,6 +68,11 @@ def test_multiline_stuff(input_path):
     )
 
 
+def test_ifabsent_stuff(input_path):
```
also great! this is the correct test. Ideally it would be named something that is a bit more informative about what it tests, like `test_enum_permissiblevalue_ifabsent` , but whatever, it is in fact testing `ifabsent_stuff`.

Silvanoc replied:
It tests 2 things: that the resulting Python code (otherwise `make_python` would fail) can be compiled and test_enum_permissiblevalue_ifabsent. That is why I left that name. I will add some comments and rename it.

sneakers-the-rat commented on linkml/generators/pythongen.py:
```diff
@@ -836,7 +930,16 @@ def gen_postinit(self, cls: ClassDefinition, slot: SlotDefinition) -> Optional[s
             else:
-                if (
+                if slot.range in self.schema.enums and slot.ifabsent:
+                    # `ifabsent` for an enumeration cannot be assigned to
```
appreciate this documentation

sneakers-the-rat commented on tests/test_generators/python/test_python_ifabsent_processor.py:
```diff
@@ -374,6 +374,44 @@ def test_uri_default_value():
+        is None
+    )
+
+    assert processor.enum_forward_references == {}
```
the test fails here because the processor doesn't have `enum_forward_references` - i think you removed that in the latest update?

Silvanoc replied:
Late hour error, will fix it.

sneakers-the-rat commented on tests/test_generators/python/test_python_ifabsent_processor.py:
```diff
@@ -374,6 +374,44 @@ def test_uri_default_value():
     )
 
 
+# If this function does not get fixture scoped to the module the
+# PythonIfAbsentProcessor instance suddenly has some attributes
+# coming from tests using the kitchen-sink example! Invalidating that
+# way the results of the test.
+@pytest.fixture(scope="module")
```
```suggestion
```

Seems to work to me except for the assert on the forward references dict! and shouldn't be a fixture anyway :)

Silvanoc replied:
Removing it, since the issue somehow magically has disappeared... And the fix was non-sense nevertheless.

sneakers-the-rat commented on linkml/generators/pythongen.py:
```diff
@@ -836,7 +930,16 @@ def gen_postinit(self, cls: ClassDefinition, slot: SlotDefinition) -> Optional[s
+                    # mutable. `python_ifabsent_processor.py` can specify
+                    # the default as string and here that string gets
+                    # converted into an object attribute invocation
+                    rlines.append(f"\tenum_name, pv_name = self.{aliased_slot_name}.split('.')")
+                    rlines.append("\tenum_cls = globals().get(enum_name)")
+                    rlines.append(f"\tself.{aliased_slot_name} = getattr(enum_cls, pv_name)")
```
OK i like the string approach, and as usual am truly not trying to be a nitpicker here, but here's an example that fails because names on enums aren't sanitized:

```yaml
CordialnessEnum:
  permissible_values:
    # ...
    ch.aotic:
      description: doesn't obey conventions

# ...

IfAbsent:
  attributes:
    ifabsent_dotted_literal:
      range: CordialnessEnum
      ifabsent: CordialnessEnum(ch.aotic)
  
```

Then we get 

```python
@dataclass(repr=False)
class IfAbsent(YAMLRoot):
    # ....
    ifabsent_dotted_literal: Optional[Union[str, "CordialnessEnum"]] = 'CordialnessEnum.ch.aotic'

    def __post_init__(self):
        # ...
        if self.ifabsent_dotted_literal is not None and not isinstance(self.ifabsent_dotted_literal, CordialnessEnum):
            enum_name, pv_name = self.ifabsent_dotted_literal.split('.')
            enum_cls = globals().get(enum_name)
            self.ifabsent_dotted_literal = getattr(enum_cls, pv_name)
```

which fails with the relatively uninformative error message: 

```
(Pdb) ksm.IfAbsent()
*** ValueError: too many values to unpack (expected 2)
```

I think that the easiest thing to do here would be to mimic the behavior of [`StrEnum`](https://docs.python.org/3/library/enum.html#enum.StrEnum) and just put the name of the enum value rather than the Enum and the value

So then that simplifies to 


```python
@dataclass(repr=False)
class IfAbsent(YAMLRoot):
    # ....
    ifabsent_dotted_literal: Optional[Union[str, "CordialnessEnum"]] = 'ch.aotic'

    def __post_init__(self):
        # ...
        if self.ifabsent_dotted_literal is not None and not isinstance(self.ifabsent_dotted_literal, CordialnessEnum):
            self.ifabsent_dotted_literal = CordialnessEnum(self.ifabsent_dotted_literal)
```

which, remarkably, works.

```python
(Pdb) ksm.CordialnessEnum('ch.aotic')
CordialnessEnum(text='ch.aotic', description="doesn't obey convention")
```

Silvanoc replied:
Good catch! You are fully right, we do not need the enum name, since the range already constraints it. So only passing the permissible value suffice.

Silvanoc replied:
BTW ensuring that corner cases are properly covered is IMO not nitpicking, but high quality code 😉

sneakers-the-rat reviewed (commented):
Great! I like the string approach a lot.

There's one change that i think should be made (removing the `@pytest.fixture` from the test in the ifabsent tests and fixing the check on the missing forward references dict), and I think you could simplify it further by just using the name of the enum value rather than the whole `Enum.value` and parsing the string, but other than that i like what ya done and think it is good.

vincentkelleher commented:
LGTM 👍 

@Silvanoc you put in a lot of effort on this one, thank you very much 🙏 

I see that you have the same Kroki.io issues as I have... 😮‍💨 I saw you've added a discussion about it too.

<img width="1061" alt="image" src="https://github.com/user-attachments/assets/2b9edf98-e23f-4895-a0f9-095a1cd23e8b">

Silvanoc commented:
I suppose that kroki.io is limiting the calls somehow and some of the tests triggered after rebasing on current `main` are now failing 🤦

sneakers-the-rat reviewed (approved):
hell ya sorry i missed that you did this three whole days ago this rocks thanks for putting up with my garbage this is great!!!!!!!!!!!

Silvanoc commented:
If I am not supposed to merge #2373 straight away to get rid of the RCOD (Red Cross of Death)[^1] in this PR, could please someone change the permissions so that I can't?! The temptation is killing me slowly...

[^1]: The small brother of the [Windows BSOD](https://www.techtarget.com/searchwindowsserver/definition/blue-screen-of-death-BSOD).

Silvanoc commented:
Now it should be ready for merging!
